Tons of build issues with missing field initializers

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks! -->

**Describe the bug**

When building with -Werror, I'm getting literally hundreds of errors related to missing designated field initializers:

```
.../test/drivers/test_suites/syscall_exit_suite/sendto_x.cpp:18:81: error: missing initializer for member ‘send_data::greater_snaplen’ [-Werror=missing-field-initializers]
   18 |         evt_test->client_to_server_ipv4_tcp(send_data{.syscall_num = __NR_sendto});
```

AFAIK partial initialization with designated initializers isn't valid C++ (which doesn't make any sense but here we are).

**How to reproduce it**

Hmm, just build with -DCMAKE_BUILD_WARNINGS_AS_ERRORS=On, I suppose.

**Expected behaviour**

I expected a successful build.

**Screenshots**

I'm not pasting all the errors here, it's literally hundreds and hundreds.

**Environment**

```
$ gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Additional context**

I'm hoping we can find some compiler-flag based solution (a gcc extension maybe), because otherwise we're looking at an insane number of patches like this one:

```
 	/*=============================== TRIGGER SYSCALL  ===========================*/
 
-	evt_test->client_to_server_ipv4_tcp(send_data{.syscall_num = __NR_sendto, .greater_snaplen = true},
-					    recv_data{.syscall_num = __NR_read});
+	evt_test->client_to_server_ipv4_tcp(
+	        send_data{.syscall_num = __NR_sendto, .greater_snaplen = true, .null_sockaddr = false},
+	        recv_data{.syscall_num = __NR_read,
+	                  .null_sockaddr = false,
+	                  .null_receiver_buffer = false,
+	                  .skip_recv_phase = false});

```

Switching to C++20 just causes *more* errors. Don't ask me why or how (this file is C, not C++) but then it also complains about https://github.com/falcosecurity/libs/blob/master/userspace/libscap/linux/scap_cgroup.c#L796 mixing designated and non-designated initializers (that one could just be `{}` I guess).

Also, am I the only one in the world building with -Werror or is my compiler sufficiently different from whatever is used in CI?

cc @Andreagit97 @federico-sysdig

--- a/.github/workflows/drivers_ci.yml
+++ b/.github/workflows/drivers_ci.yml
@@ -85,7 +85,7 @@ jobs:
         if: needs.paths-filter.outputs.driver == 'true' || needs.paths-filter.outputs.libscap == 'true' || needs.paths-filter.outputs.libpman == 'true'
         run: |
           mkdir -p build
-          cd build && cmake -DUSE_BUNDLED_DEPS=On -DBUILD_DRIVER=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DBUILD_BPF=On -DBUILD_LIBSCAP_GVISOR=${{ matrix.enable_gvisor }} -DCREATE_TEST_TARGETS=On -DENABLE_LIBSCAP_TESTS=On ../
+          cd build && cmake -DBUILD_WARNINGS_AS_ERRORS=On -DUSE_BUNDLED_DEPS=On -DBUILD_DRIVER=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DBUILD_BPF=On -DBUILD_LIBSCAP_GVISOR=${{ matrix.enable_gvisor }} -DCREATE_TEST_TARGETS=On -DENABLE_LIBSCAP_TESTS=On ../
           make scap-open driver bpf libscap_test -j6
 
       - name: Run scap-open with modern bpf 🏎️
@@ -155,7 +155,7 @@ jobs:
         if: needs.paths-filter.outputs.driver == 'true' || needs.paths-filter.outputs.libscap == 'true' || needs.paths-filter.outputs.libpman == 'true'
         run: |
           mkdir -p build
-          cd build && cmake -DUSE_BUNDLED_DEPS=ON -DENABLE_DRIVERS_TESTS=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DBUILD_BPF=True -DBUILD_LIBSCAP_GVISOR=OFF ../
+          cd build && cmake -DBUILD_WARNINGS_AS_ERRORS=On -DUSE_BUNDLED_DEPS=ON -DENABLE_DRIVERS_TESTS=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DBUILD_BPF=True -DBUILD_LIBSCAP_GVISOR=OFF ../
           make drivers_test driver bpf -j6
 
       - name: Run drivers_test with modern bpf 🏎️
@@ -202,7 +202,7 @@ jobs:
             git clone -b $GIT_BRANCH $GITHUB_SERVER_URL/$GITHUB_REPOSITORY.git libs
             cd libs
             mkdir -p build
-            cd build && cmake -DUSE_BUNDLED_DEPS=ON -DENABLE_DRIVERS_TESTS=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DBUILD_BPF=True -DBUILD_LIBSCAP_GVISOR=OFF ../
+            cd build && cmake -DBUILD_WARNINGS_AS_ERRORS=On -DUSE_BUNDLED_DEPS=ON -DENABLE_DRIVERS_TESTS=ON -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DBUILD_BPF=True -DBUILD_LIBSCAP_GVISOR=OFF ../
             make drivers_test driver bpf -j6
             sudo ./test/drivers/drivers_test -m
             rc_modern=$?
@@ -246,7 +246,7 @@ jobs:
             git config --global --add safe.directory $GITHUB_WORKSPACE
             .github/install-deps.sh
             mkdir -p build
-            cd build && cmake -DBUILD_BPF=On -DUSE_BUNDLED_DEPS=OFF -DMODERN_PROBE_INCLUDE="-I/usr/include/s390x-linux-gnu" -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DENABLE_DRIVERS_TESTS=On -DCREATE_TEST_TARGETS=On -DBUILD_LIBSCAP_GVISOR=OFF ../
+            cd build && cmake -DBUILD_WARNINGS_AS_ERRORS=On -DBUILD_BPF=On -DUSE_BUNDLED_DEPS=OFF -DMODERN_PROBE_INCLUDE="-I/usr/include/s390x-linux-gnu" -DBUILD_LIBSCAP_MODERN_BPF=ON -DMODERN_BPF_DEBUG_MODE=ON -DENABLE_DRIVERS_TESTS=On -DCREATE_TEST_TARGETS=On -DBUILD_LIBSCAP_GVISOR=OFF ../
             KERNELDIR=/lib/modules/$(ls /lib/modules)/build make driver bpf drivers_test -j6
   
   build-modern-bpf-skeleton:

--- a/test/drivers/event_class/event_class.h
+++ b/test/drivers/event_class/event_class.h
@@ -36,15 +36,15 @@ struct fd_poll {
 
 struct send_data {
 	int syscall_num;
-	bool greater_snaplen;
-	bool null_sockaddr;
+	bool greater_snaplen = false;
+	bool null_sockaddr = false;
 };
 
 struct recv_data {
 	int syscall_num;
-	bool null_sockaddr;
-	bool null_receiver_buffer;
-	bool skip_recv_phase;
+	bool null_sockaddr = false;
+	bool null_receiver_buffer = false;
+	bool skip_recv_phase = false;
 };
 
 enum protocol_L4 {
@@ -387,19 +387,23 @@ public:
 
 	void client_to_server(send_data send_d, recv_data receive_d, network_config net_config);
 	void client_to_server_ipv4_tcp(send_data send_d,
-	                               recv_data receive_d = {.skip_recv_phase = true},
+	                               recv_data receive_d = {.syscall_num = 0,
+	                                                      .skip_recv_phase = true},
 	                               int32_t client_port = IP_PORT_CLIENT,
 	                               int32_t server_port = IP_PORT_SERVER);
 	void client_to_server_ipv4_udp(send_data send_d,
-	                               recv_data receive_d = {.skip_recv_phase = true},
+	                               recv_data receive_d = {.syscall_num = 0,
+	                                                      .skip_recv_phase = true},
 	                               int32_t client_port = IP_PORT_CLIENT,
 	                               int32_t server_port = IP_PORT_SERVER);
 	void client_to_server_ipv6_tcp(send_data send_d,
-	                               recv_data receive_d = {.skip_recv_phase = true},
+	                               recv_data receive_d = {.syscall_num = 0,
+	                                                      .skip_recv_phase = true},
 	                               int32_t client_port = IP_PORT_CLIENT,
 	                               int32_t server_port = IP_PORT_SERVER);
 	void client_to_server_ipv6_udp(send_data send_d,
-	                               recv_data receive_d = {.skip_recv_phase = true},
+	                               recv_data receive_d = {.syscall_num = 0,
+	                                                      .skip_recv_phase = true},
 	                               int32_t client_port = IP_PORT_CLIENT,
 	                               int32_t server_port = IP_PORT_SERVER);
 

--- a/test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp
+++ b/test/drivers/test_suites/syscall_exit_suite/execveat_x.cpp
@@ -235,6 +235,7 @@ TEST(SyscallExit, execveatX_correct_exit) {
 	                            "2_ARGUMENT=no",
 	                            NULL};
 
+#if defined(__s390x__) || defined(__riscv) || defined(__powerpc64__) || defined(__loongarch64)
 	bool expect_truncated = true;
 	if(evt_test->is_kmod_engine() && getpagesize() > 4096) {
 		// for kmod, the size limit is actually PAGE_SIZE;
@@ -242,6 +243,7 @@ TEST(SyscallExit, execveatX_correct_exit) {
 		// In case PAGE_SIZE is < 4096, expect NON-truncated args/envs
 		expect_truncated = false;
 	}
+#endif
 
 	int flags = 0;
 

--- a/test/drivers/test_suites/syscall_exit_suite/sendto_x.cpp
+++ b/test/drivers/test_suites/syscall_exit_suite/sendto_x.cpp
@@ -141,7 +141,7 @@ TEST(SyscallExit, sendtoX_ipv4_tcp_message_not_truncated_DNS_snaplen) {
 	// The remote port is the DNS one so the snaplen should be increased.
 	evt_test->client_to_server_ipv4_tcp(
 	        send_data{.syscall_num = __NR_sendto, .greater_snaplen = true},
-	        recv_data{.skip_recv_phase = true},
+	        recv_data{.syscall_num = 0, .skip_recv_phase = true},
 	        IP_PORT_CLIENT,
 	        IP_PORT_DNS);
 

